Add support for qcow delete/discard/TRIM

djs55 opened the pull request:
This PR adds support for `delete` (aka `discard` or `TRIM`) via the Mirage Qcow library.

The capability `candelete` (used to set `ATA_SUPPORT_RZAT | ATA_SUPPORT_DRAT` in `src/lib/pci_ahci.c`) is returned from an expanded `mirage_block_stat` call, while the `delete` operation itself is plumbed through a new function `mirage_block_delete`.

Since the Mirage Qcow library supports multiple configuration options (and more are going to be added over time), an extra configuration string is plumbed through via the new block parameter `qcow-config=`.

Related to [docker/for-mac#371]

ijc25 commented:
> As the Mirage change is not yet released upstream (although I'm working on it) I've temporarily added back an opam pin to a specific branch.

I can't see that here, just the name change from `qcow-format` to `qcow`.

When updating `circle.yml` please also update the for-human steps in `README.md`.

djs55 commented:
Sorry I rebased this morning after @yallop merged the `qcow.0.6.0` upstream so I removed the pin but forgot to update the text.

rneugeba commented on src/lib/block_if.c:
```diff
@@ -427,23 +459,11 @@ blockif_proc(struct blockif_ctxt *bc, struct blockif_elem *be, uint8_t *buf)
-#endif
+		if (block_delete(bc, br->br_offset, br->br_resid) < 0) {
+			err = errno;
+			break;
```
I'm in two minds about this. In general, I don't like the these very long case statement and prefer the approach you have taken (calling a function). It does however make it harder to diff against upstream bhyve...

@ijc25 any preference?

ijc25 replied:
Looks like we already took the approach with flush of pulling it out so we may as well do so here too, but I agree wrt diffing against bhyve.

I think the `// off_t arg[2];` at the top of `blockif_proc` can/should now be removed though since this moves the vestigial reference to it elsewhere.

djs55 replied:
OK I've kept the function but have removed the `// off_t arg[2]`. CI is happy again (not that I expected it to object!)

ijc25 commented on src/lib/mirage_block_c.c:
```diff
@@ -86,29 +87,31 @@ ocaml_mirage_block_stat(mirage_block_handle h, struct stat *out, int *err) {
-		out->st_blksize = (blksize_t)sector_size;
-		out->st_flags = 0;
-		out->st_gen = 0;
+		bzero(out->stat, sizeof(struct mirage_block_stat));
```
Or another alternative would be to not embed the `stat` into the `mirage_block_stat` at all and simply add a new argument to this function for the mirage specific stuff.

ijc25 commented:
Please could you rebase onto master which now contains #82 so we can be sure this still at least builds both with and without ocaml.

djs55 commented:
OK I've
- rebased: the CI looks happy
- changed `mirage_block_stat` to have an extra parameter parameter
- fixed the indenting around the stray `else`
- replaced `qcow-format` with `qcow` in the README.md and ran through it as a sanity check
- smoke tested it again just in case :)

I think the only outstanding question is @rneugeba 's about moving the `BOP_DELETE` handling into a `block_delete` function rather than leaving it inline.

ijc25 commented:
LGTM.
